File: android/src/main/java/com/cantoo/capacitor/onnx/InferenceService.kt
package com.cantoo.capacitor.onnx

import ai.onnxruntime.NodeInfo
import ai.onnxruntime.OnnxJavaType
import ai.onnxruntime.OnnxTensor
import ai.onnxruntime.OrtEnvironment
import ai.onnxruntime.TensorInfo
import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import java.nio.ByteBuffer
import java.nio.FloatBuffer
import java.nio.IntBuffer
import java.nio.LongBuffer
import java.util.concurrent.ConcurrentHashMap

data class RawTensorInternal(
    val data: DoubleArray,
    val dims: LongArray,
    val type: String,
)

class InferenceService(
    private val sessionManager: SessionManager,
) {
    private val perSessionLock = ConcurrentHashMap<String, Mutex>()

    suspend fun prepareModel(
        modelId: String,
        version: String,
        filePath: String,
        sessionConfig: SessionConfig,
    ): String = withContext(Dispatchers.IO) {
        val session = sessionManager.ensureSession(modelId, version, filePath, sessionConfig)
        session.executionProviderUsed
    }

    suspend fun warmup(
        modelId: String,
        version: String,
        warmupInputs: Map<String, RawTensorInternal>,
    ) = withContext(Dispatchers.Default) {
        val sessionRef = sessionManager.getSession(modelId, version)
            ?: throw IllegalStateException("SESSION_INIT_ERROR: model not loaded, call loadModel first")
        if (warmupInputs.isEmpty()) return@withContext

        val lock = perSessionLock.computeIfAbsent("$modelId::$version") { Mutex() }
        lock.withLock {
            val env = sessionManager.ortEnvironment()
            val tensors = HashMap<String, OnnxTensor>()
            try {
                for ((name, raw) in warmupInputs) {
                    tensors[name] = buildOnnxTensor(env, name, raw)
                }
                sessionRef.session.run(tensors).use { /* discard output */ }
            } catch (_: Throwable) {
                // best-effort warmup; surface failures via subsequent run() calls
            } finally {
                tensors.values.forEach { it.close() }
            }
        }
    }

    suspend fun run(
        modelId: String,
        version: String,
        inputs: Map<String, RawTensorInternal>,
    ): Map<String, RawTensorInternal> = withContext(Dispatchers.Default) {
        // TEMPORARY diagnostic instrumentation — revert after debugging the Android freeze.
        Log.i("CapacitorOnnxDbg", "IS.run() enter t=${Thread.currentThread().name}")
        val sessionRef = sessionManager.getSession(modelId, version)
            ?: throw IllegalStateException("SESSION_INIT_ERROR: model not loaded, call loadModel first")
        val lock = perSessionLock.computeIfAbsent("$modelId::$version") { Mutex() }

        Log.i("CapacitorOnnxDbg", "IS.run() waiting for lock, isLocked=${lock.isLocked}")
        lock.withLock {
            Log.i("CapacitorOnnxDbg", "IS.run() LOCK ACQUIRED t=${Thread.currentThread().name}")
            if (inputs.isEmpty()) {
                throw IllegalStateException("INFERENCE_ERROR: run() requires at least one input tensor")
            }

            val session = sessionRef.session
            val inputInfo = session.inputInfo
            val env = sessionManager.ortEnvironment()
            val tensors = HashMap<String, OnnxTensor>()
            try {
                for ((name, raw) in inputs) {
                    val info = inputInfo[name]
                        ?: throw IllegalStateException("INFERENCE_ERROR: model has no input named '$name'")
                    validateTensor(name, raw, info)
                    tensors[name] = buildOnnxTensor(env, name, raw)
                }

                Log.i("CapacitorOnnxDbg", "IS.run() before session.run()")
                session.run(tensors).use { result ->
                    Log.i("CapacitorOnnxDbg", "IS.run() session.run() RETURNED")
                    val outputs = LinkedHashMap<String, RawTensorInternal>()
                    for (entry in result) {
                        val tensor = entry.value as? OnnxTensor ?: continue
                        outputs[entry.key] = readOnnxTensor(tensor)
                    }
                    if (outputs.isEmpty()) {
                        throw IllegalStateException("MODEL_INVALID: model produced no outputs")
                    }
                    Log.i("CapacitorOnnxDbg", "IS.run() outputs read, returning")
                    return@withLock outputs
                }
            } finally {
                tensors.values.forEach { it.close() }
            }
        }
    }

    private fun validateTensor(name: String, raw: RawTensorInternal, info: NodeInfo) {
        if (raw.data.isEmpty()) {
            throw IllegalStateException("INFERENCE_ERROR: input '$name' data is empty")
        }
        if (raw.dims.isEmpty()) {
            throw IllegalStateException("INFERENCE_ERROR: input '$name' dims is empty")
        }
        var elementCount = 1L
        for (dim in raw.dims) {
            if (dim <= 0L) {
                throw IllegalStateException("INFERENCE_ERROR: input '$name' dims must have positive dimensions")
            }
            elementCount *= dim
        }
        if (elementCount != raw.data.size.toLong()) {
            throw IllegalStateException("INFERENCE_ERROR: input '$name' data size does not match dims")
        }
        val tensorInfo = info.info as? TensorInfo
            ?: throw IllegalStateException("MODEL_INVALID: model input '$name' is not a tensor")
        val modelShape = tensorInfo.shape
        if (modelShape.size != raw.dims.size) {
            throw IllegalStateException("INFERENCE_ERROR: input '$name' rank does not match model input rank")
        }
        for (i in modelShape.indices) {
            if (modelShape[i] > 0L && modelShape[i] != raw.dims[i]) {
                throw IllegalStateException("INFERENCE_ERROR: input '$name' dims are incompatible with model input")
            }
        }
    }

    private fun buildOnnxTensor(
        env: OrtEnvironment,
        name: String,
        raw: RawTensorInternal,
    ): OnnxTensor {
        return when (raw.type) {
            "float32" -> {
                val arr = FloatArray(raw.data.size) { raw.data[it].toFloat() }
                OnnxTensor.createTensor(env, FloatBuffer.wrap(arr), raw.dims)
            }
            "int32" -> {
                val arr = IntArray(raw.data.size) { raw.data[it].toInt() }
                OnnxTensor.createTensor(env, IntBuffer.wrap(arr), raw.dims)
            }
            "int64" -> {
                val arr = LongArray(raw.data.size) { raw.data[it].toLong() }
                OnnxTensor.createTensor(env, LongBuffer.wrap(arr), raw.dims)
            }
            "uint8" -> {
                val arr = ByteArray(raw.data.size) { raw.data[it].toInt().toByte() }
                OnnxTensor.createTensor(env, ByteBuffer.wrap(arr), raw.dims, OnnxJavaType.UINT8)
            }
            "bool" -> {
                val arr = ByteArray(raw.data.size) { if (raw.data[it] != 0.0) 1 else 0 }
                OnnxTensor.createTensor(env, ByteBuffer.wrap(arr), raw.dims, OnnxJavaType.BOOL)
            }
            else -> throw IllegalStateException(
                "INFERENCE_ERROR: unsupported tensor type '${raw.type}' for input '$name'",
            )
        }
    }

    private fun readOnnxTensor(tensor: OnnxTensor): RawTensorInternal {
        val info = tensor.info
        val type = when (info.type) {
            OnnxJavaType.FLOAT -> "float32"
            OnnxJavaType.INT32 -> "int32"
            OnnxJavaType.INT64 -> "int64"
            OnnxJavaType.UINT8 -> "uint8"
            OnnxJavaType.BOOL -> "bool"
            else -> throw IllegalStateException(
                "INFERENCE_ERROR: unsupported output tensor type ${info.type}",
            )
        }
        return RawTensorInternal(
            data = flattenToDoubleArray(tensor.value),
            dims = info.shape,
            type = type,
        )
    }

    private fun flattenToDoubleArray(value: Any?): DoubleArray {
        val output = ArrayList<Double>(1024)

        fun visit(node: Any?) {
            when (node) {
                null -> Unit
                is Float -> output.add(node.toDouble())
                is Double -> output.add(node)
                is Int -> output.add(node.toDouble())
                is Long -> output.add(node.toDouble())
                is Byte -> output.add(node.toDouble())
                is Boolean -> output.add(if (node) 1.0 else 0.0)
                is FloatArray -> node.forEach { output.add(it.toDouble()) }
                is DoubleArray -> node.forEach { output.add(it) }
                is IntArray -> node.forEach { output.add(it.toDouble()) }
                is LongArray -> node.forEach { output.add(it.toDouble()) }
                is ByteArray -> node.forEach { output.add(it.toDouble()) }
                is BooleanArray -> node.forEach { output.add(if (it) 1.0 else 0.0) }
                is Array<*> -> node.forEach { visit(it) }
                else -> throw IllegalStateException(
                    "INFERENCE_ERROR: unsupported output type ${node::class.java.name}",
                )
            }
        }

        visit(value)
        return output.toDoubleArray()
    }
}

File: ios/Plugin/InferenceService.swift
import Foundation
// The ONNX Runtime Objective-C API ships as the `onnxruntime` module via SPM
// (onnxruntime-swift-package-manager) and as `onnxruntime_objc` via CocoaPods
// (the onnxruntime-objc pod). Resolve whichever channel the host app uses.
#if canImport(onnxruntime)
import onnxruntime
#else
import onnxruntime_objc
#endif

enum OnnxPluginError: Error {
    case modelInvalid(String)
    case sessionInitError(String)
    case inferenceError(String)
    case internalError(String)

    var code: String {
        switch self {
        case .modelInvalid: return "MODEL_INVALID"
        case .sessionInitError: return "SESSION_INIT_ERROR"
        case .inferenceError: return "INFERENCE_ERROR"
        case .internalError: return "INTERNAL_ERROR"
        }
    }

    var message: String {
        switch self {
        case .modelInvalid(let m), .sessionInitError(let m),
             .inferenceError(let m), .internalError(let m):
            return m
        }
    }

    var retryable: Bool {
        switch self {
        case .sessionInitError: return true
        default: return false
        }
    }
}

struct RawTensorInternal {
    let data: [Double]
    let dims: [Int64]
    let type: String
}

class InferenceService {
    private let sessionManager: SessionManager
    private var perSessionLocks: [String: NSLock] = [:]
    private let locksLock = NSLock()

    init(sessionManager: SessionManager) {
        self.sessionManager = sessionManager
    }

    func prepareModel(
        modelId: String,
        version: String,
        filePath: String,
        sessionConfig: SessionConfig
    ) throws -> String {
        let sessionRef = try sessionManager.ensureSession(
            modelId: modelId,
            version: version,
            filePath: filePath,
            config: sessionConfig
        )
        return sessionRef.executionProviderUsed
    }

    func warmup(modelId: String, version: String, warmupInputs: [String: RawTensorInternal]) {
        guard let sessionRef = sessionManager.getSession(modelId: modelId, version: version) else {
            return
        }
        guard !warmupInputs.isEmpty else { return }
        do {
            var feeds: [String: ORTValue] = [:]
            for (name, raw) in warmupInputs {
                feeds[name] = try buildOrtValue(name: name, raw: raw)
            }
            let outputNames = Set(try sessionRef.session.outputNames())
            _ = try sessionRef.session.run(withInputs: feeds, outputNames: outputNames, runOptions: nil)
        } catch {
            // best-effort warmup; surface failures via subsequent run() calls
        }
    }

    func run(
        modelId: String,
        version: String,
        inputs: [String: RawTensorInternal]
    ) throws -> [String: RawTensorInternal] {
        guard !inputs.isEmpty else {
            throw OnnxPluginError.inferenceError("run() requires at least one input tensor")
        }
        guard let sessionRef = sessionManager.getSession(modelId: modelId, version: version) else {
            throw OnnxPluginError.sessionInitError("model not loaded, call loadModel first")
        }

        let lock = sessionLock(for: modelId, version: version)
        lock.lock()
        defer { lock.unlock() }

        let session = sessionRef.session
        let modelInputNames = Set(try session.inputNames())
        let outputNames = try session.outputNames()
        guard !outputNames.isEmpty else {
            throw OnnxPluginError.modelInvalid("model has no outputs")
        }

        var feeds: [String: ORTValue] = [:]
        for (name, raw) in inputs {
            guard modelInputNames.contains(name) else {
                throw OnnxPluginError.inferenceError("model has no input named '\(name)'")
            }
            try validateTensor(name: name, raw: raw)
            feeds[name] = try buildOrtValue(name: name, raw: raw)
        }

        let runOutputs = try session.run(
            withInputs: feeds,
            outputNames: Set(outputNames),
            runOptions: nil
        )

        var result: [String: RawTensorInternal] = [:]
        for name in outputNames {
            guard let value = runOutputs[name] else { continue }
            result[name] = try readOrtValue(value)
        }
        guard !result.isEmpty else {
            throw OnnxPluginError.modelInvalid("model produced no output")
        }
        return result
    }

    private func validateTensor(name: String, raw: RawTensorInternal) throws {
        guard !raw.data.isEmpty else {
            throw OnnxPluginError.inferenceError("input '\(name)' data is empty")
        }
        guard !raw.dims.isEmpty else {
            throw OnnxPluginError.inferenceError("input '\(name)' dims is empty")
        }
        guard raw.dims.allSatisfy({ $0 > 0 }) else {
            throw OnnxPluginError.inferenceError("input '\(name)' dims must have positive dimensions")
        }
        let elementCount = raw.dims.reduce(1, *)
        guard elementCount == Int64(raw.data.count) else {
            throw OnnxPluginError.inferenceError("input '\(name)' data size does not match dims")
        }
    }

    private func buildOrtValue(name: String, raw: RawTensorInternal) throws -> ORTValue {
        let shape = raw.dims.map { NSNumber(value: $0) }
        let data = NSMutableData()
        let elementType: ORTTensorElementDataType

        switch raw.type {
        case "float32":
            var values = raw.data.map { Float($0) }
            data.append(&values, length: values.count * MemoryLayout<Float>.stride)
            elementType = .float
        case "int32":
            var values = raw.data.map { Int32($0) }
            data.append(&values, length: values.count * MemoryLayout<Int32>.stride)
            elementType = .int32
        case "int64":
            var values = raw.data.map { Int64($0) }
            data.append(&values, length: values.count * MemoryLayout<Int64>.stride)
            elementType = .int64
        case "uint8":
            var values = raw.data.map { UInt8($0) }
            data.append(&values, length: values.count * MemoryLayout<UInt8>.stride)
            elementType = .uInt8
        default:
            // Note: onnxruntime-objc's ORTTensorElementDataType has no `bool`
            // case, so `bool` tensors are not supported on iOS (they are on
            // Android/Web). `float16`/`uint32` are likewise unsupported here.
            throw OnnxPluginError.inferenceError(
                "unsupported tensor type '\(raw.type)' for input '\(name)'"
            )
        }

        return try ORTValue(tensorData: data, elementType: elementType, shape: shape)
    }

    private func readOrtValue(_ value: ORTValue) throws -> RawTensorInternal {
        let info = try value.tensorTypeAndShapeInfo()
        let dims = info.shape.map { $0.int64Value }
        let raw = try value.tensorData() as Data

        let data: [Double]
        let type: String
        switch info.elementType {
        case .float:
            data = raw.withUnsafeBytes { Array($0.bindMemory(to: Float.self)) }.map { Double($0) }
            type = "float32"
        case .int32:
            data = raw.withUnsafeBytes { Array($0.bindMemory(to: Int32.self)) }.map { Double($0) }
            type = "int32"
        case .int64:
            data = raw.withUnsafeBytes { Array($0.bindMemory(to: Int64.self)) }.map { Double($0) }
            type = "int64"
        case .uInt8:
            data = raw.withUnsafeBytes { Array($0.bindMemory(to: UInt8.self)) }.map { Double($0) }
            type = "uint8"
        default:
            throw OnnxPluginError.inferenceError("unsupported output tensor type")
        }

        return RawTensorInternal(data: data, dims: dims, type: type)
    }

    private func sessionLock(for modelId: String, version: String) -> NSLock {
        let key = "\(modelId)::\(version)"
        locksLock.lock()
        defer { locksLock.unlock() }
        if let existing = perSessionLocks[key] { return existing }
        let newLock = NSLock()
        perSessionLocks[key] = newLock
        return newLock
    }
}
